File: portofino-war/src/main/webapp/openid-selector/generate-sprite.js
/*
	Simple OpenID Plugin
	http://code.google.com/p/openid-selector/
	
	This code is licensed under the New BSD License.
*/

function exec(cmd) {
	var shell = new ActiveXObject('WScript.Shell');
	var exec = shell.Exec(cmd);
	while (exec.Status == 0) {
		WScript.Sleep(100);
	}
}

var imagemagick = 'C:/Program Files/ImageMagick-6.6.5-Q16/';

var locale = 'en';
if (WScript.Arguments.length == 0) {
	// assuming english locale
} else {
	locale = WScript.Arguments(0);
}

var fso = new ActiveXObject('Scripting.FileSystemObject');

var s;
var f = fso.OpenTextFile('js/openid-' + locale + '.js');
try {
	s = f.ReadAll();
} finally {
	f.Close();
}
var openid = {};
eval(s);

if (openid.locale != locale) {
	WScript.Echo('error: locale setting mismatch in js/openid-' + locale + '.js');
	WScript.Quit(-1);
}
if (openid.locale != openid.sprite) {
	WScript.Echo('error: ' + openid.locale + ' localization is reusing sprite from ' + openid.sprite 
			+ ' localization. refresh sprite on original localization or don\'t reuse the sprite');
	WScript.Quit(-1);
}

// generate small montage
var cmd = imagemagick + 'montage';
var i = 0;
for (provider_id in providers_large) {
	cmd += ' images.small/' + provider_id + '.ico.png';
	i++;
}
for (provider_id in providers_small) {
	cmd += ' images.small/' + provider_id + '.ico.png';
	i++;
}
var small = fso.GetTempName() + '.bmp';
cmd += ' -tile ' + i + 'x1 -geometry 16x16+4+4 ' + small;
exec(cmd);

// generate large montage
cmd = imagemagick + 'montage';
i = 0;
for (provider_id in providers_large) {
	cmd += ' images.large/' + provider_id + '.gif';
	i++;
}
var large = fso.GetTempName() + '.bmp';
cmd += ' -tile ' + i + 'x1 -geometry 100x60>+0+0 ' + large;
exec(cmd);

// generate final montage
var cmd = imagemagick + 'convert ' + large + ' ' + small + ' -append images/openid-providers-' + locale + '.png';
exec(cmd);

fso.DeleteFile(large);
fso.DeleteFile(small);
WScript.Echo("done");

File: portofino-pageactions/src/main/resources/META-INF/resources/openid-selector/generate-sprite.rb
#!/usr/bin/env ruby

#	Simple OpenID Plugin
#	http://code.google.com/p/openid-selector/
#	
#	This code is licenced under the New BSD License.
#
#
# Requires rubygems (tested with 1.3.7) therubyracer (tested with 0.7.5),
# and ImageMagick (tested with 6.6.5).
# Adjust path to ruby in the first line above

require 'rubygems'
require 'v8'

imagemagick = ""  # substitute specific path if needed
lang = ARGV[0] || "en"

cxt = V8::Context.new
cxt['openid'] = {}
cxt.load "js/openid-#{lang}.js"

# generate small montage
cmd = "#{imagemagick}montage"
i = 0
cxt['providers_large'].each do |provider_id, details|
  cmd += " images.small/#{provider_id}.ico.png"
  i += 1
end
cxt['providers_small'].each do |provider_id, details|
  cmd += " images.small/#{provider_id}.ico.png"
  i += 1
end
cmd += " -tile #{i}x1 -geometry '16x16+4+4' temp_small.bmp"
`#{cmd}`

# generate large montage
cmd = "#{imagemagick}montage"
i = 0
cxt['providers_large'].each do |provider_id, details|
  cmd += " images.large/#{provider_id}.gif"
  i += 1
end
cmd += " -tile #{i}x1 -geometry '100x60>+0+0' temp_large.bmp"
`#{cmd}`

# generate final montage
cmd = "#{imagemagick}convert temp_large.bmp temp_small.bmp -append images/openid-providers-#{lang}.png"
`#{cmd}`

# clean up
`rm temp_large.bmp temp_small.bmp`

nil